source/services/relay.md

```diff
@@ -22,126 +23,141 @@ Our target user hold crypto in a centralized exchange (or on another Ethereum ad
 
 Clients should verify the server’s response with the following process:
-1. Verify that the `s` that you provided in the request matches the `s` returned in the response.
-1. Verify that `tx.data` matches the bytecode of the Gnosis Safe Proxy contract with the correct owners and threshold. The `value` and `nonce` of the transaction should be zero.
-1. Hash the transaction object (`tx`) and recover the account with the provided `signature`. The resulting address should match the `deployer`.
-1. Compute the Safe address with the `deployer` address and `nonce=0`. The resulting address should match the provided `safe` address.
-
-If all checks pass, then the transaction and Safe address are valid and the user can transfer at least the `payment` amount of ETH (if `paymentToken` is address `0x0`) or the corresponding amount of `paymentToken` tokens (if the `paymentToken` is a valid token address) to the Safe address. Please take a look at the [/tokens/](relay.html#tokens-get) endpoint to see which tokens are accepted for payment by our service.
+1. Verify that `paymentReceiver` is *NULL_ADDRESS*
+2. Verify that `proxyFactory` and `masterCopy` addresses are right
+3. Call [calculateCreateProxyWithNonceAddress](https://github.com/gnosis/safe-contracts/blob/master/contracts/proxies/ProxyFactory.sol#L94) in the
+[ProxyFactory](https://github.com/gnosis/safe-contracts/blob/master/contracts/proxies/ProxyFactory.sol) using the `saltNonce` provided in the request, `masterCopy` and `setupData`
+as initializer to check that the returned address should be equal to `safe`.
+4. Decode `setupData` using [GnosisSafe](https://github.com/gnosis/safe-contracts/blob/master/contracts/GnosisSafe.sol#L76) `setup` method and check:
+   - `fallbackHandler` address is right.
+   - `owners` and `threshold` match the ones provided.
+   - `payment` is not abusive and `paymentReceiver` is *NULL_ADDRESS*
+   - `paymentToken` should be the one provided or *NULL_ADDRESS* in case it was not
+   - `to` and `data` must be empty in the current implementation. `to` and `data` could be used, e.g., to give permission to a `Safe Module` to access your funds.
+
+If all checks pass, then Safe address is valid and the user can transfer at least the `payment` amount of ETH (if `paymentToken` is address `0x0`) or the corresponding amount of `paymentToken` tokens (if the `paymentToken` is a valid token address) to the Safe address. Please take a look at the [api/v1/tokens/?gas=True](relay.html#tokens-get) endpoint to see which tokens are accepted for payment by our service.
+
+Otherwise, the response is not valid or it is compromised, and it should not be used any further.
 
-Otherwise, the response has error or it is compromised, and it should not be used any further.
-
-----------
+In the context of the Gnosis Safe and our relay services:
+- The `MasterCopy`, `ProxyFactory` and `FallbackHandler` should be addresses of a valid deployment of the `GnosisSafe` contracts.
```

Comment:
Is there somewhere a mention of the supported master copies,proxcy factoryies and fallbackhanders? if not, should we add it?

Reply:
I added one here https://github.com/gnosis/safe-docs/pull/23/files#diff-b3e28734e1cd5ff2616bc1444cb004b1R71, but yes, it won't harm to put that more visible